### ISSUE TITLE
`KubernetesClusterNodePool` reconciles although nothing changed, reconciliation fails

### ISSUE THREAD

**Issue by lippertmarkus** [crossplane-contrib/provider-jet-azure#139]
### What happened?

Directly after provisioning a new `KubernetesClusterNodePool` resource, the provider tries to reconcile it altough there were no changes to the resource. The reconciliation fails with the following error:
```
cannot run plan: plan failed: Instance cannot be destroyed: Resource azurerm_kubernetes_cluster_node_pool.ppi-demo-win1 has lifecycle.prevent_destroy set, but the plan calls for this resource to be destroyed. To avoid this error and continue with the plan, either disable lifecycle.prevent_destroy or reduce the scope of the plan using the -target flag.: File name: main.tf.json
```

Looking at the logs of the provider:
```
2022-02-14T08:19:21.707Z DEBUG provider-jet-azure Reconciling {"controller": "managed/containerservice.azure.jet.crossplane.io/v1alpha2, kind=kubernetesclusternodepool", "request": "/ppi-demo-win1"}
2022-02-14T08:19:28.669Z DEBUG provider-jet-azure refresh ended {"workspace": "/tmp/4879a21b-a78b-4968-9a93-08f76746d3a7", "out": "{\"@level\":\"info\",\"@message\":\"Terraform 1.0.5\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2022-02-14T08:19:21.728337Z\",\"terraform\":\"1.0.5\",\"type\":\"version\",\"ui\":\"0.1.0\"}\n{\"@level\":\"info\",\"@message\":\"azurerm_kubernetes_cluster_node_pool.ppi-demo-win1: Refreshing state... [id=/subscriptions/XXXX-XXXX-XXXX-XXXX-XXXX/resourcegroups/ppi-demo-aks/providers/Microsoft.ContainerService/managedClusters/cluster/agentPools/win1]\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2022-02-14T08:19:27.757622Z\",\"hook\":{\"resource\":{\"addr\":\"azurerm_kubernetes_cluster_node_pool.ppi-demo-win1\",\"module\":\"\",\"resource\":\"azurerm_kubernetes_cluster_node_pool.ppi-demo-win1\",\"implied_provider\":\"azurerm\",\"resource_type\":\"azurerm_kubernetes_cluster_node_pool\",\"resource_name\":\"ppi-demo-win1\",\"resource_key\":null},\"id_key\":\"id\",\"id_value\":\"/subscriptions/XXXX-XXXX-XXXX-XXXX-XXXX/resourcegroups/ppi-demo-aks/providers/Microsoft.ContainerService/managedClusters/cluster/agentPools/win1\"},\"type\":\"refresh_start\"}\n{\"@level\":\"info\",\"@message\":\"azurerm_kubernetes_cluster_node_pool.ppi-demo-win1: Refresh complete [id=/subscriptions/XXXX-XXXX-XXXX-XXXX-XXXX/resourcegroups/ppi-demo-aks/providers/Microsoft.ContainerService/managedClusters/cluster/agentPools/win1]\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2022-02-14T08:19:28.647525Z\",\"hook\":{\"resource\":{\"addr\":\"azurerm_kubernetes_cluster_node_pool.ppi-demo-win1\",\"module\":\"\",\"resource\":\"azurerm_kubernetes_cluster_node_pool.ppi-demo-win1\",\"implied_provider\":\"azurerm\",\"resource_type\":\"azurerm_kubernetes_cluster_node_pool\",\"resource_name\":\"ppi-demo-win1\",\"resource_key\":null},\"id_key\":\"id\",\"id_value\":\"/subscriptions/XXXX-XXXX-XXXX-XXXX-XXXX/resourcegroups/ppi-demo-aks/providers/Microsoft.ContainerService/managedClusters/cluster/agentPools/win1\"},\"type\":\"refresh_complete\"}\n{\"@level\":\"info\",\"@message\":\"Plan: 0 to add, 0 to change, 0 to destroy.\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2022-02-14T08:19:28.652519Z\",\"changes\":{\"add\":0,\"change\":0,\"remove\":0,\"operation\":\"plan\"},\"type\":\"change_summary\"}\n{\"@level\":\"info\",\"@message\":\"Apply complete! Resources: 0 added, 0 changed, 0 destroyed.\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2022-02-14T08:19:28.666877Z\",\"changes\":{\"add\":0,\"change\":0,\"remove\":0,\"operation\":\"apply\"},\"type\":\"change_summary\"}\n{\"@level\":\"info\",\"@message\":\"Outputs: 0\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2022-02-14T08:19:28.666912Z\",\"outputs\":{},\"type\":\"outputs\"}\n"}
2022-02-14T08:19:36.102Z DEBUG provider-jet-azure plan ended {"workspace": "/tmp/4879a21b-a78b-4968-9a93-08f76746d3a7", "out": "{\"@level\":\"info\",\"@message\":\"Terraform 1.0.5\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2022-02-14T08:19:28.706234Z\",\"terraform\":\"1.0.5\",\"type\":\"version\",\"ui\":\"0.1.0\"}\n{\"@level\":\"error\",\"@message\":\"Error: Instance cannot be destroyed\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2022-02-14T08:19:36.094018Z\",\"diagnostic\":{\"severity\":\"error\",\"summary\":\"Instance cannot be destroyed\",\"detail\":\"Resource azurerm_kubernetes_cluster_node_pool.ppi-demo-win1 has lifecycle.prevent_destroy set, but the plan calls for this resource to be destroyed. To avoid this error and continue with the plan, either disable lifecycle.prevent_destroy or reduce the scope of the plan using the -target flag.\",\"range\":{\"filename\":\"main.tf.json\",\"start\":{\"line\":1,\"column\":200,\"byte\":199},\"end\":{\"line\":1,\"column\":201,\"byte\":200}},\"snippet\":{\"context\":\"resource.azurerm_kubernetes_cluster_node_pool\",\"code\":\"{\\\"provider\\\":{\\\"azurerm\\\":{\\\"features\\\":{},\\\"skip_provider_registration\\\":true,\\\"subscription_id\\\":\\\"XXXX-XXXX-XXXX-XXXX-XXXX\\\"}},\\\"resource\\\":{\\\"azurerm_kubernetes_cluster_node_pool\\\":{\\\"ppi-demo-win1\\\":{\\\"availability_zones\\\":[\\\"1\\\",\\\"2\\\",\\\"3\\\"],\\\"kubernetes_cluster_id\\\":\\\"/subscriptions/XXXX-XXXX-XXXX-XXXX-XXXX/resourcegroups/ppi-demo-aks/providers/Microsoft.ContainerService/managedClusters/cluster\\\",\\\"lifecycle\\\":{\\\"prevent_destroy\\\":true},\\\"name\\\":\\\"win1\\\",\\\"node_count\\\":1,\\\"orchestrator_version\\\":\\\"1.22.4\\\",\\\"os_type\\\":\\\"Windows\\\",\\\"vm_size\\\":\\\"Standard_D4s_v5\\\"}}},\\\"terraform\\\":{\\\"required_providers\\\":{\\\"azurerm\\\":{\\\"source\\\":\\\"hashicorp/azurerm\\\",\\\"version\\\":\\\"2.78.0\\\"}}}}\",\"start_line\":1,\"highlight_start_offset\":199,\"highlight_end_offset\":200,\"values\":[]}},\"type\":\"diagnostic\"}\n"}
2022-02-14T08:19:36.102Z DEBUG provider-jet-azure Cannot observe external resource {"controller": "managed/containerservice.azure.jet.crossplane.io/v1alpha2, kind=kubernetesclusternodepool", "request": "/ppi-demo-win1", "uid": "4879a21b-a78b-4968-9a93-08f76746d3a7", "version": "10173152", "external-name": "win1", "error": "cannot run plan: plan failed: Instance cannot be destroyed: Resource azurerm_kubernetes_cluster_node_pool.ppi-demo-win1 has lifecycle.prevent_destroy set, but the plan calls for this resource to be destroyed. To avoid this error and continue with the plan, either disable lifecycle.prevent_destroy or reduce the scope of the plan using the -target flag.: File name: main.tf.json", "errorVerbose": "plan failed: Instance cannot be destroyed: Resource azurerm_kubernetes_cluster_node_pool.ppi-demo-win1 has lifecycle.prevent_destroy set, but the plan calls for this resource to be destroyed. To avoid this error and continue with the plan, either disable lifecycle.prevent_destroy or reduce the scope of the plan using the -target flag.: File name: main.tf.json\ncannot run plan\ngithub.com/crossplane/terrajet/pkg/controller.(*external).Observe\n\t/root/provider-jet-azure/.work/pkg/pkg/mod/github.com/crossplane/terrajet@v0.3.1/pkg/controller/external.go:184\ngithub.com/crossplane/crossplane-runtime/pkg/reconciler/managed.(*Reconciler).Reconcile\n\t/root/provider-jet-azure/.work/pkg/pkg/mod/github.com/crossplane/crossplane-runtime@v0.15.1-0.20211004150827-579c1833b513/pkg/reconciler/managed/reconciler.go:681\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/root/provider-jet-azure/.work/pkg/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.6/pkg/internal/controller/controller.go:298\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/root/provider-jet-azure/.work/pkg/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.6/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/root/provider-jet-azure/.work/pkg/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.6/pkg/internal/controller/controller.go:214\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1581"}
2022-02-14T08:19:36.103Z DEBUG controller-runtime.manager.events Warning {"object": {"kind":"KubernetesClusterNodePool","name":"ppi-demo-win1","uid":"4879a21b-a78b-4968-9a93-08f76746d3a7","apiVersion":"containerservice.azure.jet.crossplane.io/v1alpha2","resourceVersion":"10173152"}, "reason": "CannotObserveExternalResource", "message": "cannot run plan: plan failed: Instance cannot be destroyed: Resource azurerm_kubernetes_cluster_node_pool.ppi-demo-win1 has lifecycle.prevent_destroy set, but the plan calls for this resource to be destroyed. To avoid this error and continue with the plan, either disable lifecycle.prevent_destroy or reduce the scope of the plan using the -target flag.: File name: main.tf.json"}
```

What's interesting is that the `refresh ended` log line says `Plan: 0 to add, 0 to change, 0 to destroy.` and also `Apply complete! Resources: 0 added, 0 changed, 0 destroyed.`.

The `plan ended` log line however, then states that the resource needs to be removed.

### How can we reproduce it?
```yaml
apiVersion: azure.jet.crossplane.io/v1alpha2
kind: ResourceGroup
metadata:
  annotations:
    crossplane.io/external-name: ppi-demo-aks
  name: ppi-demo
spec:
  forProvider:
    location: germanywestcentral
    tags:
      provisioner: crossplane
  providerConfigRef:
    name: ppi-test
---
apiVersion: network.azure.jet.crossplane.io/v1alpha2
kind: VirtualNetwork
metadata:
  annotations:
    crossplane.io/external-name: vnet
  name: ppi-demo
spec:
  forProvider:
    addressSpace:
      - 192.168.0.0/16
    location: germanywestcentral
    resourceGroupNameRef:
      name: ppi-demo
  providerConfigRef:
    name: ppi-test
---
apiVersion: network.azure.jet.crossplane.io/v1alpha2
kind: Subnet
metadata:
  annotations:
    crossplane.io/external-name: subnet
  name: ppi-demo
spec:
  forProvider:
    addressPrefixes:
      - 192.168.0.0/17
    resourceGroupNameRef:
      name: ppi-demo
    virtualNetworkNameRef:
      name: ppi-demo
  providerConfigRef:
    name: ppi-test
---
apiVersion: containerservice.azure.jet.crossplane.io/v1alpha2
kind: KubernetesCluster
metadata:
  annotations:
    crossplane.io/external-name: cluster
  name: ppi-demo
spec:
  forProvider:
    defaultNodePool:
      - availabilityZones:
          - '1'
          - '2'
          - '3'
        name: default
        nodeCount: 1
        orchestratorVersion: 1.22.4
        vmSize: Standard_D2s_v5
        vnetSubnetIdRef:
          name: ppi-demo
    dnsPrefix: ppi-demo
    identity:
      - type: SystemAssigned
    kubernetesVersion: 1.22.4
    location: germanywestcentral
    networkProfile:
      - loadBalancerSku: Standard
        networkPlugin: azure
        networkPolicy: calico
    resourceGroupNameRef:
      name: ppi-demo
    roleBasedAccessControl:
      - enabled: true
    skuTier: Free
    windowsProfile:
      - adminPasswordSecretRef:
          key: password
          name: ppi-demo
          namespace: clusters-internal-ppi-demo
        adminUsername: vmadmin
  providerConfigRef:
    name: ppi-test
  writeConnectionSecretToRef:
    name: ppi-demo-kubeconfig
    namespace: crossplane
---
apiVersion: containerservice.azure.jet.crossplane.io/v1alpha2
kind: KubernetesClusterNodePool
metadata:
  annotations:
    crossplane.io/external-name: win1
  name: ppi-demo-win1
spec:
  forProvider:
    availabilityZones:
      - '1'
      - '2'
      - '3'
    kubernetesClusterIdRef:
      name: ppi-demo
    nodeCount: 1
    orchestratorVersion: 1.22.4
    osType: Windows
    vmSize: Standard_D4s_v5
  providerConfigRef:
    name: ppi-test
```

### What environment did it happen in?
Crossplane version: v1.6.2
provider-jet-azure: v0.7.0-preview

**Comment by ulucinar** [comment]
A related upstream issue is here:
https://github.com/hashicorp/terraform-provider-azurerm/issues/8297

**Comment by ulucinar** [comment]
My observation is that although `spec.forProvider` does not have a `vnet_subnet_id` argument, the resulting `terraform.tfstate` contains it. And a subsequent `terraform plan` triggers a destroy because it detects a change in the `vnet_subnet_id`, i.e.,
```
 - vnet_subnet_id         = "/subscriptions/<subscription id>/resourceGroups/alper/providers/Microsoft.Network/virtualNetworks/vnet-alper/subnets/subnet-alper" -> null # forces replacement
```

Btw, I could also reproduce this issue using the following HCL configuration:

```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "=2.78.0"
    }
  }
}


provider "azurerm" {
  features {}
}

resource "azurerm_kubernetes_cluster_node_pool" "ppi-demo-win1" {
  name                  = "winalp"
  kubernetes_cluster_id = "/subscriptions/<subscription id>/resourcegroups/alper/providers/Microsoft.ContainerService/managedClusters/cluster-alper"
  vm_size               = "Standard_D2_v2"
  node_count            = 1
  orchestrator_version  = "1.22.4"
  os_type               = "Windows"
  availability_zones    = ["1", "2", "3"]
  lifecycle {
    prevent_destroy     = true
  }
}
```

**Comment by ulucinar** [comment]
I've proposed a [change](https://github.com/crossplane/terrajet/pull/263) in terrajet that helps us with this issue. As I explained in that PR's description, we can categorize this issue as an upstream Terraform issue and, as a workaround, always [require](https://github.com/hashicorp/terraform-provider-azurerm/issues/8297#issuecomment-683697089) that we set the `spec.forProvider.VnetSubnetID`. However, the UX will be degraded and we will have unnecessary redundancy, the Kubernetes cluster with which a nodepool is associated can already determine the subnet for the pool, and hence there should be no need to explicitly specify it with a separate field (duplication). Thus, I proposed a change in how we late-initialize in terrajet.
